### PR TITLE
Cristian/cc3200 v3.0

### PR DESCRIPTION
**Pull request by crisap94**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,234 +21,106 @@ The SimpleLink™ Wi-Fi® CC3200 LaunchPad™ development kit (with QFN-packaged
 
 **IMPORTANT NOTE**: To be able to compile the CC3200 you have establish the [jumpper settings](http://energia.nu/cc3200guide/) properly.
```
Comment (jgarciade) [inline README.md]: you have to **set**

File: README.md [hunk 2]
```diff
@@ -22,234 +21,106 @@ The SimpleLink™ Wi-Fi® CC3200 LaunchPad™ development kit (with QFN-packaged
...
+Ubidots(const char* token, UbiServer server, IotProtocol iot_protocol)
+```
+
+> @token, [Required]. Your Ubidots unique account [TOKEN](http://help.ubidots.com/user-guides/find-your-token-from-your-ubidots-account).  
+> @server, [Optional], [Options] = [`UBI_INDUSTRIAL`, `UBI_EDUCATIONAL`], [Default] = `UBI_INDUSTRIAL`. The server to send data, set `UBI_EDUCATIONAL` if your account is educational type.  
+> @iot_protocol, [Optional], [Options] = [`UBI_HTTP`, `UBI_TCP`, `UBI_UDP`], [Default] = `UBI_TCP`. The IoT protocol that you will use to send or retrieve data.
+
+Creates an Ubidots instance.
+
+**NOTE:** If you use HTTP or TCP to send data, the client will implement TLS 2.0 using the ![native ESP8266 secure client](https://github.com/esp8266/Arduino/blob/master/doc/esp8266wifi/client-secure-examples.rst) to secure your data. Keep in mind that due to the security checks needed, the packet may take a little more time to be sent than without TLS. If you wish to send data insecurely, use UDP or any of the available examples at our ![docs](https://ubidots.com/docs/hw/).
```
Comment (jgarciade) [inline README.md]: this is not true, this library does not implement TLS

File: README.md [hunk 3]
```diff
@@ -22,234 +21,106 @@ The SimpleLink™ Wi-Fi® CC3200 LaunchPad™ development kit (with QFN-packaged
...
+
+As Ubidots makes its best to secure your data, we do not guarantee any issue, data miss or external sniff coming from the native secure client or bugs in the library.
```
Comment (jgarciade) [inline README.md]: delete this

File: src/UbiBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+Developed and maintained by Jose Garcia for Ubidots Inc
+@jotathebest at github: https://github.com/jotathebest
+*/
+
+#include "UbiBuilder.h"
+
+#include "UbiHttp.h"
+
+#include "UbiTcp.h"
+
```
Comment (jgarciade) [inline src/UbiBuilder.cpp]: delete the new line spaces

File: src/UbiBuilder.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+@jotathebest at github: https://github.com/jotathebest
+*/
+
+#ifndef _UbiBuilder_H_
+#define _UbiBuilder_H_
+
+#include "Arduino.h"
+
+#include "UbiProtocol.h"
+
```
Comment (jgarciade) [inline src/UbiBuilder.h]: idem

File: src/UbiConstants.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+@jotathebest at github: https://github.com/jotathebest
+*/
+
+#ifndef _UbiConstants_H_
+#define _UbiConstants_H_
+
+#include "UbiTypes.h"
+#include "stdint.h"
+
+const char* const UBIDOTS_SERVER = "industrial.api.ubidots.com";
```
Comment (jgarciade) [inline src/UbiConstants.h]: this server is repeated

File: src/UbiConstants.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+const int UBIDOTS_HTTP_PORT = 80;
+const int UBIDOTS_HTTPS_PORT = 443;
+const int UBIDOTS_TCP_PORT = 9012;
+const int UBIDOTS_TCPS_PORT = 9812;
+const uint8_t MAX_VALUES = 10;
+const float ERROR_VALUE = -3.4028235E+8;
+const int MAX_BUFFER_SIZE = 700;
+static UbiServer UBI_INDUSTRIAL = "industrial.api.ubidots.com";
+static UbiServer UBI_EDUCATIONAL = "things.ubidots.com";
+const int MAX_PROTOCOLS = 3;
```
Comment (jgarciade) [inline src/UbiConstants.h]: change this to SUPPORTED_PROTOCOLS

File: src/UbiHttp.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  UbiHTTP(const char *server, const int port, const char *user_agent,
+          const char *token);
+
+  ~UbiHTTP();
+
+  bool sendData(const char *device_label, const char *device_name,
+                char *payload);
+
+  float get(const char *device_label, const char *variable_label);
+
```
Comment (jgarciade) [inline src/UbiHttp.h]: delete the blank new line spaces

File: src/UbiProtocolHandler.cpp [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+/**
+ * Builds the HTTP payload to send and saves it to the input char pointer.
+ * @payload [Mandatory] char payload pointer to store the built structure.
+ * @timestamp_global [Optional] If set, it will be used for any dot without
+ * timestamp.
+ */
+
+void UbiProtocolHandler::buildHttpPayload(char* payload) {
```
Comment (jgarciade) [inline src/UbiProtocolHandler.cpp]: move both tcp and http payload builders to their respective classes. With this, we may update the other libraries

File: src/UbiProtocolHandler.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  bool _debug;
+
+  UbiProtocol* _ubiProtocol;
+  IotProtocol _iot_protocol;
+  UbiToken _token;
+  Value* _dots;
+
+  void buildHttpPayload(char* payload);
+  void buildTcpPayload(char* payload, const char* device_label, const char* device_name);
+
```
Comment (jgarciade) [inline src/UbiProtocolHandler.h]: delete blank spaces

File: src/UbiTcp.cpp [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+ */
+bool UbiTCP::sendData(const char *device_label, const char *device_name,
+                      char *payload) {
+  /* Connecting the client */
+  _client_tcp_ubi.connect(_server, UBIDOTS_TCP_PORT);
+  reconnect(_server, UBIDOTS_TCP_PORT);
+
+  /* Sends data to Ubidots */
+  if (_client_tcp_ubi.connected()) {
+    _client_tcp_ubi.print(payload);
```
Comment (jgarciade) [inline src/UbiTcp.cpp]: add a way to see the payload to send if the user sets debug to true

File: src/UbiTcp.h [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+*/
+
+#ifndef _UbiTcp_H_
+#define _UbiTcp_H_
+
+#include "UbiProtocol.h"
+#include <WiFi.h>
+#include <time.h>
+
+#pragma GCC diagnostic push
```
Comment (jgarciade) [inline src/UbiTcp.h]: this is not necessary

File: src/UbiTypes.h [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+typedef struct ContextUbi {
+
+  char *key_label;
+  char *key_value;
+  
+} ContextUbi;
+
+typedef const char *UbiServer;
+
```
Comment (jgarciade) [inline src/UbiTypes.h]: delete blank spaces

File: src/UbiUdp.h [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+  ~UbiUDP();
+
+  bool sendData(const char *device_label, const char *device_name,
+                char *payload);
+
+  float get(const char *device_label, const char *variable_label);
+
+  void setDebug(bool debug);
+  
+  bool serverConnected();
```
Comment (jgarciade) [inline src/UbiUdp.h]: delete blank spaces

File: src/Ubidots.h [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  bool send();
+  bool send(const char *device_label);
+  bool send(const char *device_label, const char *device_name);
+
+  float get(const char *device_label, const char *variable_label);
+
+  void setDebug(bool debug);
+
+  bool wifiConnect(const char *ssid, const char *password);
+  bool wifiConnected();
```
Comment (jgarciade) [inline src/Ubidots.h]: add a method to set fixed ip, dns and network mask, something like:

wifiConnectSettings( ip, mask, dns);

this method must be invoked before wifiConnect().

Create an additional example showing how to use it and also update the docs

**Comment by crisap94** [comment]
Everything is updated as requested. Doing some tests

**Pull request merged**